lucene/core/src/java/org/apache/lucene/codecs/lucene99/GroupVintReader.java

```diff
@@ -0,0 +1,176 @@
+    {2, 1, 1, 0}, {2, 1, 1, 1}, {2, 1, 1, 2}, {2, 1, 1, 3},
+    {2, 1, 2, 0}, {2, 1, 2, 1}, {2, 1, 2, 2}, {2, 1, 2, 3},
+    {2, 1, 3, 0}, {2, 1, 3, 1}, {2, 1, 3, 2}, {2, 1, 3, 3},
+    {2, 2, 0, 0}, {2, 2, 0, 1}, {2, 2, 0, 2}, {2, 2, 0, 3},
+    {2, 2, 1, 0}, {2, 2, 1, 1}, {2, 2, 1, 2}, {2, 2, 1, 3},
+    {2, 2, 2, 0}, {2, 2, 2, 1}, {2, 2, 2, 2}, {2, 2, 2, 3},
+    {2, 2, 3, 0}, {2, 2, 3, 1}, {2, 2, 3, 2}, {2, 2, 3, 3},
+    {2, 3, 0, 0}, {2, 3, 0, 1}, {2, 3, 0, 2}, {2, 3, 0, 3},
+    {2, 3, 1, 0}, {2, 3, 1, 1}, {2, 3, 1, 2}, {2, 3, 1, 3},
+    {2, 3, 2, 0}, {2, 3, 2, 1}, {2, 3, 2, 2}, {2, 3, 2, 3},
+    {2, 3, 3, 0}, {2, 3, 3, 1}, {2, 3, 3, 2}, {2, 3, 3, 3},
+    {3, 0, 0, 0}, {3, 0, 0, 1}, {3, 0, 0, 2}, {3, 0, 0, 3},
+    {3, 0, 1, 0}, {3, 0, 1, 1}, {3, 0, 1, 2}, {3, 0, 1, 3},
+    {3, 0, 2, 0}, {3, 0, 2, 1}, {3, 0, 2, 2}, {3, 0, 2, 3},
+    {3, 0, 3, 0}, {3, 0, 3, 1}, {3, 0, 3, 2}, {3, 0, 3, 3},
+    {3, 1, 0, 0}, {3, 1, 0, 1}, {3, 1, 0, 2}, {3, 1, 0, 3},
+    {3, 1, 1, 0}, {3, 1, 1, 1}, {3, 1, 1, 2}, {3, 1, 1, 3},
+    {3, 1, 2, 0}, {3, 1, 2, 1}, {3, 1, 2, 2}, {3, 1, 2, 3},
+    {3, 1, 3, 0}, {3, 1, 3, 1}, {3, 1, 3, 2}, {3, 1, 3, 3},
+    {3, 2, 0, 0}, {3, 2, 0, 1}, {3, 2, 0, 2}, {3, 2, 0, 3},
+    {3, 2, 1, 0}, {3, 2, 1, 1}, {3, 2, 1, 2}, {3, 2, 1, 3},
+    {3, 2, 2, 0}, {3, 2, 2, 1}, {3, 2, 2, 2}, {3, 2, 2, 3},
+    {3, 2, 3, 0}, {3, 2, 3, 1}, {3, 2, 3, 2}, {3, 2, 3, 3},
+    {3, 3, 0, 0}, {3, 3, 0, 1}, {3, 3, 0, 2}, {3, 3, 0, 3},
+    {3, 3, 1, 0}, {3, 3, 1, 1}, {3, 3, 1, 2}, {3, 3, 1, 3},
+    {3, 3, 2, 0}, {3, 3, 2, 1}, {3, 3, 2, 2}, {3, 3, 2, 3},
+    {3, 3, 3, 0}, {3, 3, 3, 1}, {3, 3, 3, 2}, {3, 3, 3, 3}
+  };
+
+  IndexInput in;
+  private static final int[] MASKS = {0xFF, 0xFFFF, 0xFFFFFF, 0xFFFFFFFF};
+
+  // buffer for all groups
+  private int offset = 0;
+  private byte[] bytes = new byte[16];
+
+  private byte numGroup = 0;
+  private int[] groupLengths = null;
+
+  // the next int will be read in the single group. in the range [0-3].
+  private int posInGroup = 0;
+
+  public GroupVintReader() {}
+
+  /** Called when decode a new block. */
+  public void reset(IndexInput indexInput, int num) throws IOException {
+    this.in = indexInput;
+    offset = 0;
+    posInGroup = 0;
+
+    if (num > 0) {
+      int len = in.readVInt();
+      if (len > 0) {
+        numGroup = in.readByte();
+        // + 3 bytes to avoid BitUtil.VH_LE_INT.get out of array bounds when reading the last value
+        bytes = ArrayUtil.growNoCopy(bytes, len + 3);
+        in.readBytes(bytes, 0, len);
+      }
+    }
+  }
+
+  private void readVInts(long[] docs, int off, int limit) throws IOException {
+    for (int i = off; i < limit; i++) {
+      docs[i] = in.readVInt();
+    }
+  }
+
+  /** only readValues or nextInt can be called after reset */
+  public void readValues(long[] docs, int limit) throws IOException {
+    if (numGroup == 0) {
+      readVInts(docs, 0, limit);
+      return;
+    }
+    int groupValues = limit / 4 * 4;
```

Comment:
We can do this with a single instruction I believe?

```suggestion
    int groupValues = limit & 0xFFFFFFFC;
```

Reply:
nice idea :)